data-pipeline/telemetry-location-updater/src/main/java/org/ekstep/ep/samza/service/TelemetryLocationUpdaterService.java

```diff
@@ -10,42 +11,128 @@
+	private Path path = new Path();
+	private Type mapType = new TypeToken<Map<String, Object>>() { }.getType();
 
-
-	public TelemetryLocationUpdaterService(DeviceProfileCache deviceLocationCache, JobMetrics metrics) {
+	public TelemetryLocationUpdaterService(DeviceProfileCache deviceLocationCache, JobMetrics metrics, RedisConnect redisConnect, Config config) {
 		this.deviceProfileCache = deviceLocationCache;
 		this.metrics = metrics;
+		this.userStoreDb = config.getInt("redis.database.userStore.id", 4);
+		this.redisConnect = redisConnect;
+		this.userDataStoreConnection = this.redisConnect.getConnection(userStoreDb);
 	}
 
 	public void process(TelemetryLocationUpdaterSource source, TelemetryLocationUpdaterSink sink) {
 		try {
 			Event event = source.getEvent();
-			// Add device location details to the event
-			updateEventWithIPLocation(event);
-			sink.toSuccessTopic(event);
+			String did = event.did();
+			DeviceProfile deviceProfile = null;
+			if (did != null && !did.isEmpty()) {
+				// check for user profile
+				Map<String, String> derivedLocation = getLocationFromUserCache(event);
+				// get device profile from cache
+				deviceProfile = deviceProfileCache.getDeviceProfileForDeviceId(did);
+
+				// get user declared location if user profile is empty
+				if (derivedLocation.isEmpty()) {
+					derivedLocation = getUserDeclaredLocation(deviceProfile);
+				}
+				// get ip reso lved location if user declared location is empty
+				if (derivedLocation.isEmpty()) {
+					derivedLocation = getIpResolvedLocation(deviceProfile);
+				}
+				// Add derived location to telemetry
+				event.addDerivedLocation(derivedLocation);
+
+				// Add device profile details to the event
+				updateEvent(event, deviceProfile);
+				metrics.incProcessedMessageCount();
+				sink.toSuccessTopic(event);
+			}
+			else {
+				updateEvent(event, deviceProfile);
+				metrics.incUnprocessedMessageCount();
+				sink.toSuccessTopic(event);
+			}
 		} catch (JsonSyntaxException e) {
 			LOGGER.error(null, "INVALID EVENT: " + source.getMessage());
 			sink.toMalformedTopic(source.getMessage());
 		}
 	}
 
-	private void updateEventWithIPLocation(Event event) {
-		String did = event.did();
-		DeviceProfile deviceProfile = null;
-		if (did != null && !did.isEmpty()) {
-			deviceProfile = deviceProfileCache.getDeviceProfileForDeviceId(event.did());
-			updateEvent(event, deviceProfile);
-			metrics.incProcessedMessageCount();
-		} else {
-			updateEvent(event, deviceProfile);
-			metrics.incUnprocessedMessageCount();
+	private Map<String, String> getLocationFromUserCache(Event event) {
+		String uid = event.actorid();
+		Map<String, String> locationData;
+		try {
+			locationData = getLocationForUser(uid);
+			return locationData;
+		}
+		catch (JedisException ex) {
+			LOGGER.error(null, "Reconnecting with Redis store due to exception: ", ex);
+			redisConnect.resetConnection();
+			try (Jedis redisConn = redisConnect.getConnection(userStoreDb)) {
```

Comment:
The redis get and internal retry code can be pushed to a util class as this is common for all reads from redis for all denorm jobs..